kops: Add a kops-updown job

zmerlynn opened the pull request:
This is dipping a toe in the water. Companion to https://github.com/kubernetes/kubernetes/pull/33518, and I won't merge it until that's in and a new docker image is pushed.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/test-infra/681)
<!-- Reviewable:end -->

fejta commented on jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-kops-aws.yaml:
```diff
@@ -0,0 +1,67 @@
+            {provider-env}
+            {job-env}
+            {post-env}
+            if ! curl -fsS --retry 3 -o "$WORKSPACE/kops.url" "https://storage.googleapis.com/kops-ci/bin/latest-ci.txt"; then
```
This logic seems too complicated to live inside the yaml...

Will you move this into a script and add it to the kubekins-e2e image like we do with `e2e-runner.sh`, `e2e.go`, etc?

https://github.com/kubernetes/test-infra/blob/master/jenkins/e2e-image/Dockerfile

zmerlynn replied:
The part that's really messy without doing this inline is setting the `NODEUP_URL`. I need to save both the NODEUP_URL export and the kops binary. I can do it `ssh-agent -c` style and have the binary spit out the necessary exports, but it's dirty.

zmerlynn replied:
We should talk about this. I think it defeats the purpose. To most of us writing these jobs, you've got the `yaml`, which is super transparent, you've got `e2e-runner.sh`, which is anything but transparent, and then you hit `e2e.go`. This is a super minor setup/download, all things considered, but you didn't want it in `e2e.go`, where it would be obvious and easy to maintain in `go`. It's *not* easy to maintain in `e2e-runner.sh`, which is actually the next most obvious place to put it, because I need to plumb an env variable through. It's not easy to maintain in a separate binary, because I end up writing `yaml` code that looks like this:
```
if ! kops-download.sh $WORKSPACE/kops $WORKSPACE/kops.env; then
  echo "Can't retrieve kops" >&2
  exit 1
fi
. $WORKSPACE/kops.env
```

That's about two lines less complicated than the `yaml` code I wrote already, which is only noisy because of the rigorous error checking that a *ton* of the other `yaml` shell code *doesn't do*. Furthermore, neither case is dockerized, which is kind of bad, but not the worst thing in the world. So, in conclusion, I think this should either stay in the `yaml` or move down into `e2e.go`, and not get lost in some inscrutable morass in the middle.

zmerlynn replied:
As for what I mean by error checking, this section of bash is easily rewritten as:
```
export KOPS_URL=$(curl -fsS --retry 3 https://storage.googleapis.com/kops-ci/bin/latest-ci.txt)"
curl -fsS --retry 3 -o "${KOPS_URL}/kops" "$(cat $WORKSPACE/kops.url)/linux/amd64/kops"
export NODEUP_URL="${KOPS_URL}/linux/amd64/nodeup"
```

fejta replied:
The example code you give is still too complex for yaml. Conditional logic is complicated and therefore belongs in its own script, which jenkins job can call unconditionally:

`kops-runner.sh`
```
if ! kops-download.sh $WORKSPACE/kops; then
  exit 1
fi
# Whatever complex stuff you want
. $WORKSPACE/kops.env
timeout {runner}
```

`kops-aws.yaml`
```
export KUBERNETES_PROVIDER="aws"  # Pretty soon we will disallow this complexity too
bash <(curl -fsS --retry 3  "https://raw.githubusercontent.com/kubernetes/test-infra/master/jenkins/kops-runner.sh")  # Pretty soon we will automatically and always checkout test-infra at master
```

I do not want complex logic in yaml, and I'm spending a lot of effort in this space. I allowed it initially because at least in the yaml it is source controlled, but fundamentally there needs to be an easy way for someone to run your job without jenkins.

If you encapsulate this logic inside a `kops-runner.sh` or `kops-runner.go` program it will be much easier for someone to run this job the same way jenkins will: call the program.

zmerlynn replied:
I did it a little differently than you described and just overrode the `e2e-runner.sh` in `dockerized-e2e-runner.sh`, then bundled it in the Docker image. WTDT? (Testing the normal path in https://github.com/kubernetes/kubernetes/pull/33700 right now.)

fejta replied:
:+1:

fejta commented on jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-kops-aws.yaml:
```diff
@@ -0,0 +1,67 @@
+    builders:
+        - activate-gce-service-account
+        - shell: |
+            {provider-env}
```
Any interest in saving us all some pain and place `provider-env` and `job-env` here inline?

zmerlynn replied:
I'm going to have more jobs, but moving the `provider-env` is easy enough.

fejta commented on jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-kops-aws.yaml:
```diff
@@ -0,0 +1,67 @@
+            job-env: |
+                export GINKGO_TEST_ARGS="--ginkgo.focus=\[k8s.io\]\sNetworking.*\[Conformance\]"
+                export GINKGO_PARALLEL="y"
+                export PROJECT="fake-project-ignored"
```
Hmm? Sadly I think this will cause our tooling to validate that fake-project-ignored has the right IAM bindings :)

zmerlynn replied:
Oh, I can fix `e2e-runner`. It just needs a provider check around the `PROJECT` check.

fejta commented on jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-kops-aws.yaml:
```diff
@@ -0,0 +1,67 @@
+            days-to-keep: 7
+    triggers:
+        - reverse:
+            jobs: '{trigger-job}'
```
Consider removing as many of these as appropriate (we only trigger one way, we only define one version of emails)

zmerlynn replied:
I've looked at this a little more, and this is basically stock boilerplate for any build triggered job template. If you want a wider cleanup, let's address it in another PR, but https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml and https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gke.yaml both follow this. Am I missing something?

fejta replied:
kubernetes-e2e-gce.yaml has multiple values for trigger-job:

`kubernetes-e2e-gce` uses [`kubernetes-build`](https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml#L50) while kubernetes-e2e-gce-federation uses [`kubernetes-federation-build`](https://github.com/kubernetes/test-infra/blob/master/jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-gce.yaml#L401)

Your job has only one value for trigger-job right now, therefore the intermediate variable is unnecessary.

If you write `jobs: 'kubernetes-build'` the next person has the information right in front of their eyes instead of needing to go hunt down where trigger-job is defined.

zmerlynn replied:
Done.

zmerlynn commented:
I addressed some of the review comments and fixed up `kubernetes-e2e-aws.yaml` similarly. I'm happy to discuss offline where to put the download code.

fejta commented on jenkins/dockerized-e2e-runner.sh:
```diff
@@ -61,6 +61,12 @@ if [[ "${KUBE_RUN_FROM_OUTPUT:-}" =~ ^[yY]$ ]]; then
 
+if [[ -n "${KUBE_E2E_RUNNER:-}" ]]; then
+    docker_extra_args+=(\
+      --entrypoint="${KUBE_E2E_RUNNER}" \
```
Neat!

fejta commented on jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-aws.yaml:
```diff
@@ -17,24 +17,20 @@
         - activate-gce-service-account
         - shell: |
-            {provider-env}
+            export KUBERNETES_PROVIDER="aws"
```
Yay! This looks MUCH better!!

fejta commented on jenkins/job-configs/kubernetes-jenkins/kubernetes-e2e-kops-aws.yaml:
```diff
@@ -0,0 +1,54 @@
+    builders:
+        - activate-gce-service-account
+        - shell: |
+            export AWS_CONFIG_FILE="${{WORKSPACE}}/.aws/credentials"
```
This one too :-D

fejta commented on jenkins/e2e-image/Dockerfile:
```diff
@@ -44,10 +44,11 @@ ENV FAIL_ON_GCP_RESOURCE_LEAK=true \
 # JENKINS_AWS_CREDENTIALS_FILE
 
 ADD ["e2e-runner.sh", \
+    "kops-e2e-runner.sh", \
```
Do you plan to add this file to the PR?

zmerlynn replied:
Woops. Fixed!

fejta commented:
LGTM

fejta commented:
Note that we will need to push a new kubekins-e2e image for this (talk to @ixdy)

zmerlynn commented:
Yup. I can do it, or someone else can. If this seems fine, merge and I can push a new image and suggest a new tag later (but it may not be until Monday if I'm doing it).

fejta commented:
Given that various pieces here do and do not run from head I'd suggest we wait until Monday morning when you are back to a) merge this PR and b) create the new image.

I suspect if we merge now the aws tests will spend the rest of the week in a fail loop

zmerlynn commented:
I just rebased and I'm retesting this in #33700 again. I'll merge when that seems clean, then propose a new tag.

zmerlynn commented:
Merging, #33700 is collecting greens.
